### PR TITLE
CIoU protected divides

### LINKED ISSUE #8539: Bug loss calculation: 0. in pred follows in NaNs in bbox_iou(...)
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Training, Validation

### Bug

The problem occurs when the prediction has 0.0 values. These then follow through the method bbox_iou in NANs in the loss calculation.
https://github.com/ultralytics/yolov5/blob/526e650553819dbff67897b9c752c4072e989823/train.py#L352-L353
https://github.com/ultralytics/yolov5/blob/526e650553819dbff67897b9c752c4072e989823/utils/loss.py#L138-L142
In the example below, the NaNs result here:
https://github.com/ultralytics/yolov5/blob/526e650553819dbff67897b9c752c4072e989823/utils/metrics.py#L247
This is one reason for such a training: 
results

If you agree with me, I'll make a PR :) 

### Environment

YOLOv5 torch 1.11 (cuda 11.3) and 1.12 (cuda 11.6)

### Minimal Reproducible Example

```
pxy = torch.tensor([[0.2729, 0.6113],
        [0.6572, 0.6045],
        [0.2646, 0.6406],
        [0.6221, 0.6357],
        [0.3359, 0.2222],
        [0.7158, 0.4521]], device='cuda:0', dtype=torch.float16,
       requires_grad=True)
pwh = torch.tensor([[7.3277e-08, 5.9422e-07],
        [0.0000e+00, 0.0000e+00],
        [7.5654e-08, 9.5031e-07],
        [0.0000e+00, 0.0000e+00],
        [1.5166e-06, 1.4725e-05],
        [3.2162e-11, 0.0000e+00]], device='cuda:0', requires_grad=True)
pbox = torch.cat((pxy, pwh), 1)
tbox = torch.tensor([[ 0.6705,  0.7127,  4.0370,  2.7977],
        [ 0.6705,  0.7127,  4.0370,  2.7977],
        [-0.3295,  0.7127,  4.0370,  2.7977],
        [-0.3295,  0.7127,  4.0370,  2.7977],
        [ 0.6705, -0.2873,  4.0370,  2.7977],
        [ 0.6705, -0.2873,  4.0370,  2.7977]], device='cuda:0')
iou = bbox_iou(pbox, tbox, CIoU=True).squeeze()  # iou(prediction, target)
print(iou) # tensor([-0.0711,     nan, -0.0914,     nan, -0.0851, -0.0420])
```

### Additional

One solution would be to add `torch.finfo(torch.float32).tiny`

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by glenn-jocher**
Protected divides in IOU function to resolve https://github.com/ultralytics/yolov5/issues/8539

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Optimization of bounding box intersection-over-union calculations in YOLOv5.

### 📊 Key Changes
- 📐 Removed an epsilon (`eps`) addition in the width and height calculations of bounding boxes.
- 🔍 Adjusted the `v` term in the Complete Intersection over Union (CIoU) loss calculation to prevent division by zero during the aspect ratio term computation.

### 🎯 Purpose & Impact
- 🧹 The changes serve to clean up the intersection-over-union (IoU) computation, potentially leading to more accurate bounding box evaluations.
- 👨‍💻 For users and developers, these tweaks might result in slight performance enhancements during model training and evaluation, and more precise object detection metrics.

### CODE REVIEW

**Comment by UnglvKitDe** [comment]
@glenn-jocher Looks good! What about lines 232/233 and 250? There, it is also possible to divide with 0. Isn't it?:
https://github.com/ultralytics/yolov5/blob/7dafd1cb297869032d98406afc9f3e74f68b5bcd/utils/metrics.py#L232-L233
https://github.com/ultralytics/yolov5/blob/7dafd1cb297869032d98406afc9f3e74f68b5bcd/utils/metrics.py#L250

**Comment by glenn-jocher** [comment]
@UnglvKitDe 233 is not a divide, the \ symbol means line continuation. The c2 on L250 is already protected by the `+ eps` when it is defined in L244:
https://github.com/ultralytics/yolov5/blob/7dafd1cb297869032d98406afc9f3e74f68b5bcd/utils/metrics.py#L244

**Comment by UnglvKitDe** [comment]
@glenn-jocher You are right, srry didn't read correctly.

**Review by UnglvKitDe — APPROVED** [review]

**Comment by UnglvKitDe** [comment]
@glenn-jocher Thanks for the quick fix!

**Comment by glenn-jocher** [comment]
@UnglvKitDe PR is merged. Thank you for your contributions to YOLOv5 🚀 and Vision AI ⭐
